File: class_member/4.py
class Cal(object):
    # _언더바를 사용하는 이유는 외부에서는 사용하지 않는 변수라는 뜻
    _history = []
    #생성자
    #2번째 arg가 a로 가, 그래서 맨 앞에 self를 추가
    def __init__(self,a,b):
        if isinstance(a, int):
            self.a=a
        if isinstance(b, int):
            self.b=b

    def add(self):
        result = self.a+self.b
        Cal._history.append("add : %d+%d=%d" %(self.a, self.b, result))
        return result
    def subtract(self):
        result = self.a-self.b
        Cal._history.append("subtract : %d-%d=%d" %(self.a, self.b, result))
        return result
    def setA(self, a):
        if isinstance(a,int):
            self.a=a
    def setB(self, b):
        if isinstance(b,int):
            self.b=b
    def getA(self):
        return self.a
    def getB(self):
        return self.b
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)

class CalMultiply(Cal):
    def multiply(self):
        result = self.a*self.b
        Cal._history.append("multiply : %d*%d=%d" %(self.a, self.b, result))
        return result

class CalDivide(CalMultiply):
    def divide(self):
        result = self.a/self.b
        Cal._history.append("divide : %d/%d=%d" %(self.a, self.b, result))
        return result

#initialize method with name of class
c1 = CalMultiply(10,10)
print(c1.add())
print(c1.subtract())

c2 = CalMultiply(5,3)
print(c2.multiply())

c3 = CalDivide(35,7)
print(c3.divide())

Cal.history()

File: class_member/4.rb
class Cal
  attr_reader :a, :b
  attr_writer :a
  @@_history=[]
  #생성자
  #인스턴스가 생성될 때 꼭 실행되어야 하는 메소드
  def initialize(a, b)
    @a = a  #인스턴스 변수 = 지역 변수
    @b = b
  end
  def add
    result = @a+@b
    #문자열 안에 변수를 넣고 싶으면 #{} 안에 쓰면 된다
    @@_history.push "add : #{@a}+#{@b}=#{result}"
    return result
  end
  def subtract
    result = @a-@b
    @@_history.push "subtract : #{@a}-#{@b}=#{result}"
    return result
  end
  def setA a
    if a.is_a? Integer
      @a=a
    end
  end
  def setB b
    if b.is_a? Integer
      @b=b
    end
  end
  def getA
    return @a
  end
  def getB
    return @b
  end
  def Cal.history
    for item in @@_history
      p item
    end
  end
end

class CalMultiply < Cal
  def multiply
    result = @a*@b
    @@_history.push "multiply : #{@a}*#{@b}=#{result}"
    return result
  end
end

class CalDivide < CalMultiply
  def divide
    result = @a/@b
    @@_history.push "divide : #{@a}/#{@b}=#{result}"
    return result
  end
end

#initialize method with 'new' method
c1 = Cal.new(10,10)
p c1.add
p c1.subtract

c2 = CalMultiply.new(6,4)
p c2.multiply

c3 = CalDivide.new(16,4)
p c3.divide
Cal.history
